### PR TITLE
code-gen: allow specifying custom resync periods for certain informer types

### PR DESCRIPTION
**Pull request by natronq**
**What this PR does / why we need it**:
This PR extends the informer code-generator to allow the consumer to specify a custom resync period for certain informer types and uses the default resync period if none is defined.

**Special notes for your reviewer**:
Example:
```go
cs := clientset.NewForConfigOrDie(config)
resyncConfig := externalversions.ResyncConfiguration{
  &samplev1alpha1.Sample{}: 30 * time.Second,
}
informer := externalversions.NewSharedInformerFactory(cs, 2*time.Minute, externalversions.WithCustomResyncConfig(resyncConfig))
```
**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/client/informers/externalversions/factory.go [hunk 1]
```diff
@@ -47,20 +50,27 @@ type sharedInformerFactory struct {
...
 func NewSharedInformerFactory(client versioned.Interface, defaultResync time.Duration) SharedInformerFactory {
-	return NewFilteredSharedInformerFactory(client, defaultResync, v1.NamespaceAll, nil)
+	return NewFilteredSharedInformerFactory(client, defaultResync, v1.NamespaceAll, nil, nil)
 }
 
 // NewFilteredSharedInformerFactory constructs a new instance of sharedInformerFactory.
 // Listers obtained via this SharedInformerFactory will be subject to the same filters
 // as specified here.
-func NewFilteredSharedInformerFactory(client versioned.Interface, defaultResync time.Duration, namespace string, tweakListOptions internalinterfaces.TweakListOptionsFunc) SharedInformerFactory {
+func NewFilteredSharedInformerFactory(client versioned.Interface, defaultResync time.Duration, namespace string, tweakListOptions internalinterfaces.TweakListOptionsFunc,
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/client/informers/externalversions/factory.go]: why does `NewFilteredSharedInformerFactory` get the `resyncConfig` param, but `NewSharedInformerFactory` doesn't?

Instead of adding more args, we should switch to another extension pattern, maybe `informer.WithCustomResyncs`.

Comment (natronq) [reply staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/client/informers/externalversions/factory.go]: Adding the parameter to `NewSharedInformerFactory` would cause a huge refactor accross the Kubernetes code-base, because all consumers of this factory almost exclusivley use this method to instantiate a new factory. 

Changing the arguments of `NewFilteredSharedInformerFactory` does not break any existing functionality as it is not used verbatim. So this seemed like a natural fit as it also already incorporates custom list options.

Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/client/informers/externalversions/factory.go]: What I wanted to say is that adding another param is the wrong pattern.

Comment (natronq) [reply staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/client/informers/externalversions/factory.go]: Ok, so you would suggest adding a seperate method e.g. `NewSharedInformerFactoryWithCustomResync`instead?

Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/client/informers/externalversions/factory.go]: No. Rather I would prefer some functional style pattern:

`NewSharedInformerFactory(...).WithTweakListOptions(...).WithCustomResyncConfig(...)`.

File: pkg/client/informers/informers_generated/internalversion/factory.go [hunk 2]
```diff
@@ -44,36 +44,63 @@ import (
...
+	return NewFilteredSharedInformerFactory(client, defaultResync, v1.NamespaceAll, options...)
 }
 
-// NewFilteredSharedInformerFactory constructs a new instance of sharedInformerFactory.
-// Listers obtained via this SharedInformerFactory will be subject to the same filters
-// as specified here.
-func NewFilteredSharedInformerFactory(client internalclientset.Interface, defaultResync time.Duration, namespace string, tweakListOptions internalinterfaces.TweakListOptionsFunc) SharedInformerFactory {
-	return &sharedInformerFactory{
+// NewFilteredSharedInformerFactory constructs a new instance of sharedInformerFactory contained to the specified namespace with additional options.
+func NewFilteredSharedInformerFactory(client internalclientset.Interface, defaultResync time.Duration, namespace string, options ...SharedInformerOption) SharedInformerFactory {
```
Comment (sttts) [inline pkg/client/informers/informers_generated/internalversion/factory.go]: we have to leave the old func and reimplement it with `NewInformerFactoryWithOptions`.

File: pkg/client/informers/informers_generated/internalversion/factory.go [hunk 3]
```diff
@@ -44,36 +44,63 @@ import (
 	storage "k8s.io/kubernetes/pkg/client/informers/informers_generated/internalversion/storage"
 )
 
+type ResyncConfiguration map[v1.Object]time.Duration
```
Comment (sttts) [inline pkg/client/informers/informers_generated/internalversion/factory.go]: do we really need a `ResyncConfiguration` type? Sounds "over-abstracted".

File: pkg/generated/bindata.go [hunk 4]
```diff
@@ -13651,7 +13651,7 @@ msgstr "Imposta il current-context in un file kubeconfig"
 
 #: pkg/kubectl/cmd/describe.go:86
 msgid "Show details of a specific resource or group of resources"
-msgstr "Mostra i dettagli di una specifiche risorsa o un gruppo di risorse"
+msgstr "Mostra i dettagli di una specifica risorsa o un gruppo di risorse"
```
Comment (sttts) [inline pkg/generated/bindata.go]: unexpected change. Make it another commit at least if this fix is right.

Comment (natronq) [reply pkg/generated/bindata.go]: This was not intended. I think it was caused by running update-codegen.sh against the current master.

File: pkg/client/informers/informers_generated/internalversion/factory.go (hunk 2 above)
Comment (sttts) [reply pkg/client/informers/informers_generated/internalversion/factory.go]: Disregard my previous comment. it's compatible. But still, I would not add the options here. Instead I would even deprecate it in favor of the generic other variant with options.

File: pkg/generated/bindata.go (hunk 4 above)
Comment (sttts) [reply pkg/generated/bindata.go]: Do we have a spell checker included now? :)

Comment (natronq) [reply pkg/generated/bindata.go]: Looks like it https://github.com/kubernetes/kubernetes/blob/3ea4e449c361a66eb3539433487cd22d7256b032/hack/verify-spelling.sh
:-)

Comment (sttts) [reply pkg/generated/bindata.go]: Wow. Anyway, saw the same issue. The checked in bindata is broken it seems.

File: pkg/client/informers/informers_generated/internalversion/factory.go (hunk 2 above)
Comment (natronq) [reply pkg/client/informers/informers_generated/internalversion/factory.go]: As `NewFilteredSharedInformerFactory` is not used anywhere I deleted it and introdued `NewSharedInformerFactoryWithOptions`.

File: pkg/generated/bindata.go (hunk 4 above)
Comment (natronq) [reply pkg/generated/bindata.go]: Separated it in another commit.

File: pkg/client/informers/informers_generated/internalversion/factory.go (hunk 2 above)
Comment (sttts) [reply pkg/client/informers/informers_generated/internalversion/factory.go]: It's used by 3rdparties. Would leave it and just deprecate it.

Comment (natronq) [reply pkg/client/informers/informers_generated/internalversion/factory.go]: Done.

File: pkg/client/informers/informers_generated/internalversion/factory.go [hunk 5]
```diff
@@ -44,36 +44,78 @@ import (
...
 // Listers obtained via this SharedInformerFactory will be subject to the same filters
 // as specified here.
+// Deprecated: Please use NewSharedInformerFactoryWithOptions instead
 func NewFilteredSharedInformerFactory(client internalclientset.Interface, defaultResync time.Duration, namespace string, tweakListOptions internalinterfaces.TweakListOptionsFunc) SharedInformerFactory {
-	return &sharedInformerFactory{
+	return NewSharedInformerFactoryWithOptions(client, defaultResync, WithNamespace(namespace), WithTweakListOptions(tweakListOptions))
+}
+
+// NewSharedInformerFactoryWithOptions constructs a new instance of a SharedInformerFactory with additional options.
+func NewSharedInformerFactoryWithOptions(client internalclientset.Interface, defaultResync time.Duration, options ...SharedInformerOption) SharedInformerFactory {
```
Comment (sttts) [inline pkg/client/informers/informers_generated/internalversion/factory.go]: While not exactly the same type, this is compatible with `NewSharedInformerFactory` calls. It might be worth to merge those two again. I think the chance is high that nobody passes around constructors such that this change will make any trouble.

Comment (natronq) [reply pkg/client/informers/informers_generated/internalversion/factory.go]: I agree that compatibility is probably not an issue here, but I actually prefer having a separate method for passing the options. It is more verbose and immediately clear to the caller that this method accepts additional options.

**Comment by ncdc** [comment]
Style is great. Let's do this!

File: staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go [hunk 6]
```diff
@@ -98,36 +99,78 @@ func (g *factoryGenerator) GenerateType(c *generator.Context, t *types.Type, w i
 }
 
 var sharedInformerFactoryStruct = `
+type SharedInformerOption func(*sharedInformerFactory) *sharedInformerFactory
```
Comment (ncdc) [inline staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go]: Add godoc?

File: staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go [hunk 7]
```diff
@@ -98,36 +99,78 @@ func (g *factoryGenerator) GenerateType(c *generator.Context, t *types.Type, w i
...
+	customResync map[{{.reflectType|raw}}]{{.timeDuration|raw}}
 
 	informers map[{{.reflectType|raw}}]{{.cacheSharedIndexInformer|raw}}
 	// startedInformers is used for tracking which informers have been started.
 	// This allows Start() to be called multiple times safely.
 	startedInformers map[{{.reflectType|raw}}]bool
 }
 
-// NewSharedInformerFactory constructs a new instance of sharedInformerFactory
+// Sets a custom resync period for the specified informer types.
```
Comment (ncdc) [inline staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go]: Please use the function name as the beginning of the comment, i.e. `// WithCustomResyncConfig sets a ...`

File: staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go [hunk 8]
```diff
@@ -98,36 +99,78 @@ func (g *factoryGenerator) GenerateType(c *generator.Context, t *types.Type, w i
...
+func WithCustomResyncConfig(resyncConfig map[{{.object|raw}}]{{.timeDuration|raw}}) SharedInformerOption {
+	return func(factory *sharedInformerFactory) *sharedInformerFactory {
+		for k, v := range resyncConfig {
+			factory.customResync[reflect.TypeOf(k)] = v
+		}
+		return factory
+	}
+}
+
+// Sets a custom filter on all listers of the configured SharedInformerFactory.
```
Comment (ncdc) [inline staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go]: Please use the function name as the beginning of the comment, i.e. `// WithTweakListOptions sets a ...`

File: staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go [hunk 9]
```diff
@@ -98,36 +99,78 @@ func (g *factoryGenerator) GenerateType(c *generator.Context, t *types.Type, w i
...
+func WithTweakListOptions(tweakListOptions internalinterfaces.TweakListOptionsFunc) SharedInformerOption {
+	return func(factory *sharedInformerFactory) *sharedInformerFactory {
+		factory.tweakListOptions = tweakListOptions
+		return factory
+	}
+}
+
+// Limits the SharedInformerFactory to the specified namespace.
```
Comment (ncdc) [inline staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go]: Please use the function name as the beginning of the comment, i.e. `// WithNamespace limits the ...`

**Review by ncdc — CHANGES_REQUESTED** [review]

**Comment by ncdc** [comment]
@natronq thanks for this - I really like it.

**Comment by natronq** [comment]
@ncdc Refactored according to your comments.

**Review by ncdc — APPROVED** [review]

**Comment by deads2k** [comment]
> /cc @caesarxuchao @ncdc any opinion about this style? I like this much more than adding more and more constructors. This pattern is much more idiomatic.

I dislike the style, but as long as the types remain private to make it impossible to write "trust me, I'm doing good things" functions I won't fight it.

/approve
